jib-cli/src/main/java/com/google/cloud/tools/jib/cli/jar/JarModeProcessor.java

```diff
@@ -131,6 +138,81 @@ static JarType determineJarType(Path jarPath) throws IOException {
 
+  /**
+   * Creates layers as specified in BOOT-INF/layers.idx (if present) or for dependencies,
+   * spring-boot-loader, snapshot dependencies, resource and classes for a spring boot fat jar.
+   *
+   * @param jarPath path to jar file
+   * @param localExplodedJarRoot path to temporary jib local directory
+   * @return list of {@link FileEntriesLayer}
+   * @throws IOException if I/O error occurs when opening the jar file or if temporary directory
+   *     provided doesn't exist
+   */
+  static List<FileEntriesLayer> createLayersForExplodedSpringBootFat(
+      Path jarPath, Path localExplodedJarRoot) throws IOException {
+
+    try (JarFile jarFile = new JarFile(jarPath.toFile())) {
+      ZipEntry layerIndex = jarFile.getEntry("BOOT-INF/layers.idx");
+      ZipUtil.unzip(jarPath, localExplodedJarRoot);
+      if (layerIndex != null) {
+        return createLayersForLayeredSpringBootJar(localExplodedJarRoot);
+      }
+
+      ArrayList<FileEntriesLayer> layers = new ArrayList<>();
+
+      // Non-snapshot layer
+      Predicate<Path> isInBootInfLib =
+          path -> path.getParent().startsWith(localExplodedJarRoot.resolve("BOOT-INF/lib"));
+      Predicate<Path> isSnapshot = path -> path.getFileName().toString().contains("SNAPSHOT");
+      Predicate<Path> nonSnapshotPredicate = isInBootInfLib.and(isSnapshot.negate());
+      FileEntriesLayer nonSnapshotLayer =
+          addDirectoryContentsToLayer(
+              DEPENDENCIES, localExplodedJarRoot, nonSnapshotPredicate, APP_ROOT);
+
+      // Snapshot layer
+      Predicate<Path> snapshotPredicate = isInBootInfLib.and(isSnapshot);
+      FileEntriesLayer snapshotLayer =
+          addDirectoryContentsToLayer(
+              SNAPSHOT_DEPENDENCIES, localExplodedJarRoot, snapshotPredicate, APP_ROOT);
+
+      // Spring-boot-loader layer.
+      Predicate<Path> isLoader =
+          path -> path.getParent().startsWith(localExplodedJarRoot.resolve("org"));
+      FileEntriesLayer loaderLayer =
+          addDirectoryContentsToLayer(
+              "spring-boot-loader", localExplodedJarRoot, isLoader, APP_ROOT);
+
+      // Classes layer.
+      Predicate<Path> isClass = path -> path.getFileName().toString().endsWith(".class");
+      Predicate<Path> isInBootInfClasses =
+          path ->
+              path.getParent()
+                  .startsWith(
+                      localExplodedJarRoot.resolve(
+                          localExplodedJarRoot.resolve("BOOT-INF/classes")));
+      Predicate<Path> finalPredicateClasses = isInBootInfClasses.and(isClass);
+      FileEntriesLayer classesLayer =
+          addDirectoryContentsToLayer(
+              CLASSES, localExplodedJarRoot, finalPredicateClasses, APP_ROOT);
+
+      // Resources layer.
+      Predicate<Path> isInMetaInf =
+          path -> path.getParent().startsWith(localExplodedJarRoot.resolve("META-INF"));
+      Predicate<Path> finalPredicateResources =
+          isInBootInfClasses.or(isInMetaInf).and(isClass.negate());
```

Comment:
I think the following is more accurate:
```
isInMetaInf.or(isInBootInfClasses.and(isClass.negate()))
```
And I thought we should also include resources files at the JAR content root? Is it not the case?

Reply:
Sorry, would that mean resources files outside the `BOOT-INF/classes/` and `META-INF/` directories?

Reply:
No, it means `isInMetaInf || (isInBootInfClasses && !isClass)`. OTOH, your original condition is `(isInBootInfClasses || isInMetaInf) && !isClass)`.

Reply:
That said, I'm curious if a spring jar can have resources files at root? Perhaps not?

Reply:
(OTOH, a regular jar has resource files for the application at root.)

Reply:
I wonder if it can't be like this for a spring jar:

1. resources under `BOOT-INF/classes/` --> resources for the app
1. resources under root (excluding `BOOT-INF/`) --> resources for the spring launcher? (Not accessible from the app runtime but they are for configuring Spring Boot?)
1. resources under `META-INF/` --> for the app or the launcher? Or both?

Reply:
hm afaik,[ spring boot jars are structured in this way:](https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-executable-jar-format.html#executable-jar-nested-jars) 
``` 
+-META-INF
 |  +-MANIFEST.MF
 +-org
 |  +-springframework
 |     +-boot
 |        +-loader
 |           +-<spring boot loader classes>
 +-BOOT-INF
    +-classes
    |  +-mycompany
    |     +-project
    |        +-YourClasses.class
    +-lib
       +-dependency1.jar
       +-dependency2.jar
```  
The application classes and resources go into `BOOT-INF/classes/`. `META-INF/` can also contain some files such as `pom.properties`, `pom.xml` and `build-info.properties`.  But outside of those three directories (META/, BOOT-INF/, org/) I don't see the possibility of any other resources existing - at least going by the standard structure of the spring boot JAR.